monitoring/monitoring/utils.py

```diff
@@ -0,0 +1,218 @@
+    line chart for extracted and scraped products for all timestamps found by merchant and country.
+
+    Returns:
+        List: [plotly chart, pandas dataframe, pandas dataframe]
+    """
+    extraction = all_extraction_summary()
+    extraction["type"] = "extraction"
+    scraping = all_scraping_summary()
+    scraping["type"] = "scraping"
+    all_timestamps = pd.concat([extraction, scraping], ignore_index=True)
+    all_timestamps["date"] = pd.to_datetime(all_timestamps["timestamp"]).dt.date
+    df_timestamps_by_merchant = (
+        all_timestamps.groupby(by=["date", "merchant", "type"])
+        .sum()
+        .pivot_table(
+            values="products",
+            index=["date", "merchant"],
+            columns="type",
+            aggfunc=np.sum,
+            fill_value=0,
+        )
+        .sort_values(by="date", ascending=False)
+    )
+    df_timestamps_by_merchant_country = (
+        all_timestamps.groupby(by=["date", "merchant", "country", "type"])
+        .sum()
+        .sort_values(by="date", ascending=False)
+    )
+    df = all_timestamps.groupby(by=["date", "merchant", "type"]).sum().reset_index()
+    fig_timestamps_by_merchant = px.line(
+        df, x="date", y="products", color="merchant", symbol="type"
+    )
+    return (
+        fig_timestamps_by_merchant,
+        df_timestamps_by_merchant,
+        df_timestamps_by_merchant_country,
+        all_timestamps,
+    )
+
+
+def all_timestamps_agg() -> tuple:
+    """
+    Creates an aggregated view of 'all_timestamps_merchant'. Resulting objects show total
+    number of extracted and scraped products for all timestamps found.
+
+    Returns:
+        List: [plotly chart, pandas dataframe]
+    """
+    df = all_timestamps_merchant()[3].groupby(by=["date", "type"]).sum().reset_index()
+    df_timestamps_agg = df.pivot_table(
+        values="products", index=["date"], columns="type", aggfunc=np.sum, fill_value=0
+    ).sort_values(by="date", ascending=False)
+    fig_timestamps_agg = px.line(df, x="date", y="products", color="type")
+    return fig_timestamps_agg, df_timestamps_agg
+
+
+def products_by_label() -> tuple:
+    """
+    Gets 'products_by_sustainability_label' from green-db table, counts products with
+    "certificate:UNKNOWN" and products with any other label and creates a dataframe and a line
+    chart from it. It also filters given query to present products by sustainability label(s)
+    from last product extraction in a dataframe.
+
+    Returns:
+        List: [plotly chart, pandas dataframe, pandas dataframe]
+    """
+    query = green_db.products_by_sustainability_label()
+    unknown = []
+    known = []
+    for row in query:
+        for item in row:
+            if type(item) is list:
+                for label in item:
+                    if label == "certificate:UNKNOWN":
+                        unknown.append(row)
+                    else:
+                        known.append(row)
```

Comment:
You should pull this into the API implementation and let the database do the work, which is more performant and reduces memory footprint

Reply:
Removed and filtered directly on the query to db ([6800091](https://github.com/calgo-lab/green-db/pull/91/commits/68000913abc6ced9cff53a0be8aa7df59596e658))